### ISSUE TITLE
Shared Build & Test Pipeline

### ISSUE THREAD

**Issue by jhuntoo** [CitadelJS/core#4]
We need to consider to what, if any extent can we share tooling across repos.

### Potentially shareable
- Building and npm package publishing
- webpack/ tslint / babel etc
- Unit Testing configuration
- HTTP Acceptance Testing

**Comment by otbe** [comment]
- build config (for webpack, rollup, browserify, tsd etc)
- test config (for webpack, rollup, tape, ava, karma)
- linter config (tslint..)
- general (.gitignore, .editorconfig, .npmignore....)

I'd prefer to outsource npm publishing to travis or something like that.
HTTP acceptance tests should be only necessary in @citadel/routing, shouldn't it?

**Comment by lholznagel** [comment]
Would say that we try to limitate the use of tools like webpack or gulp and use npm scripts as often as possible. I now that there can be many scripts but with a little organization we can save lots of dependencies, oh and lots of installing time

**Comment by lholznagel** [comment]
expect for webpack and browserify I agree with @otbe
Currently don´t really now for what we need rollup (seems to be more browser side) and karma is also more browser side
also we should decide tape, ava or mocha. I would tend to ava or mocha wwith chai

**Comment by otbe** [comment]
Webpack makes it very simple to use TS and babel in conjunction with superior source map support and we can run babel and TS specific tests at the same time, but I guess that's a bit of a philosophical discussion :)

+1 for mocha (should be enough for a node-only library)

**Comment by lholznagel** [comment]
sorry always when I used webpack I hated it. But you are right its a philosophical discussion :)

**Comment by jhuntoo** [comment]
@lholznagel yes you can do much with just npm scripts.
@otbe Given the level of modularity we are seeking, integration testing will be important. It's so annoying when modular frameworks release stuff where it's hard to find the "golden set" of modules that work 😠 . So I'd like to see a fair bit of testing the actual via HTTP, so integration bugs don't slip through the cracks.

In fact, maybe we could have a integrations test repo that runs tests over master of all repos. Then we could have more confidence that individually working repos aren't breaking the "bigger picture"

**Comment by jhuntoo** [comment]
+1 for mocha / chai - personal familiarity and it's seems popular for nodejs
+ 1 for webpack

**Comment by lholznagel** [comment]
Great idea
Maybe this repo can react on all commits of all repos and run the tests. Think that should be possible with travis

**Comment by lholznagel** [comment]
and with a mocha and nyc (coverage) we can run our tests without compiling to Javascript. Big advantage on time

**Comment by remojansen** [comment]
👍 webpack
👍 mocha + chai
👍 integration tests

**Comment by otbe** [comment]
Some scribblings https://github.com/otbe/inversify-express-utils/tree/webpack-build-test-coverage

**Comment by lholznagel** [comment]
Just took a quick look. Maybe create a subfolder in the tests with called babel and typescript or keep it like now but also name the typescript files like you named the babel files?

**Comment by otbe** [comment]
@lholznagel yes maybe renaming or other folder structure would be good :)
The only reason for ```framework.ts``` and ```framework_babel.js``` is that WebStorm shows files with the same name but different file extensions in a "merged" way and I does not like it :D
Webpack distinguish both file by file extension.

**Comment by lholznagel** [comment]
Ah okay. Thought that every test will run as ts and babel for example and how to use purposes.

**Comment by otbe** [comment]
Mh this would force us to duplicate a lot of tests and test time will be raised a lot. I think only critical parts have to be babel-tested (all decorators). JSDocs + TS test code + general examples should be enough for JS and TS devs. Any thoughts on this?

**Comment by lholznagel** [comment]
Yeah think you are right, to have everything duplicated is a little bit overkill

**Comment by lholznagel** [comment]
For sharing configs we could use git submodules. 
Advantages:
- Not tooling
- Directly in git
- Separate repo and versioning

Disadvantage:
- can´t insert submodule into root
 -> solution: put them into the folder config and create a symlink to root

**Comment by otbe** [comment]
Got some time today to work on https://github.com/otbe/inversify-express-utils/tree/webpack-build-test-coverage

- tsd generation works now, but we need to add a small script which removes /// references for projects that are not @types ready (like inversify :D)
- refactoring of code base especially removed old interface style
- using @types where it is possible and added type deps to package.json

```
{
    "inversify-express-utils": "git+ssh://git@github.com:otbe/inversify-express-utils.git#webpack-build-test-coverage"
}
```
Its usable and works down to node v4 without polyfills :)
Below node v4 you will need some polyfills for Symbol, WeakMap and such things but from code perspective its full useable. :)

@lholznagel 
Thought about the same idea, but symlink are problematic for some OS's and tools (will they follow symlinks? :D) Have some trouble at the past....mh...

**Comment by lholznagel** [comment]
Nice progress.
I only tested with my example repo and Linux that worked fine. Don´t know what other OS´s and tools say. But currently this is the only elegant way I could think of. Always open for something different :)

**Comment by lholznagel** [comment]
Did some testing with using git subtree. With it we would have a similar setup but without using symlinks, the files would be directly in the root of the project.

**Comment by jhuntoo** [comment]
@lholznagel Wow I've never used git subtree before, looks promising. 👍

**Comment by lholznagel** [comment]
yeah it looked promising.

Here from out gitter chat:
> okay after some tries with @otbe we found that there is a problem with existing files. When you use `git read-tree -m -u <branch>` all other files are deleted
